XLS-30: AMM Specification (#125)
XRPLF/XRPL-Standards: XLS-30/README.md

Note that in this version, only equal-weighted two asset pools are supported. However, differing weighted pools could be supported in future versions.

#### 2.1.1. Ledger Entries representing AMM instance

##### **`AccountRoot`** ledger entry

A new flag **`lsfAMM`** is introduced to the [**`AccountRoot`**](https://xrpl.org/accountroot.html) object. 

| Flag Name |  Hex Value | Description |
| ---------- | :----------------: | :-------:|

> | `lsfAMM`      | 0x40000000  |   If set, it indicates that the object represents an AMM instance|

Reviewer: `lsfAMM` is obsolete and is replaced with `AMMID`.

Author: Deleted.

  - If the computed `AccountRootID` exists, repeat
    - else, return `AccountRootID`
 
#### **`AMM`** ledger entry 
The unique ID of the new **`AMM`** object , a.k.a **`AMMID`** is computed as follows:
- Calculate the `SHA512-Half` of some of the following values:
  - The `issuer` of the issued asset;
  - The `currency` of the issued asset;
  - The `issuer` of the second issued asset, if there exists one;
  - The `currency` of the second issued asset, if there exists one;
  <!-- `XRP` if the `Balance` field exists, i.e. if one of the assets is XRP "GT: why is this needed? We just use issuer/currency of each asset. XRP defaults both to 0."-->
  
The order of the fields to compute the hash is decided canonically. The **`AMMID`** associated with this **`AccountRootID`** is this hash to ensure the uniqueness of the AMM instance. The applications can look-up the **`AccountRootID`** for a specific AMM instance by providing the asset pair for that instance.

The **`AMM`** ledger entry contains the following fields for the **`AccountRoot`** object that represents the AMM instance.

> 1. **`AccountID`** specifies the ID of the `AccountRoot` object associated with this `AMM` ledger entry.

Reviewer: It's `Account`, not `AccountID`.

Author: Done.

2. **`TradingFee`** specifies the fee, in basis point, to be charged to the traders for the trades executed against this AMM instance. Valid values for this field are between 0 and 1000 inclusive. A value of 1 is equivalent to 1/10 bps or 0.001%, allowing trading fee between 0% and 1%. Trading fee is a percentage of the trade size. It is charged on the asset being deposited for the `AMMDeposit` (if applicable) and `Payment` transactions and on the asset being withdrawan for the `AMMWithdraw` (if applicable) transaction. This fee is added to the AMM instance's pools and is distributed to the LPs in proportion to the `LPTokens` upon liquidity removal.

3. **`VoteSlots`** represents an array of `Vote` objects.
4. **`AuctionSlot`** represents the `Auction` object.
5. **`LPTokenBalance`** specifies the balance of outstanding liquidity Provider Tokens (LPTokens).
6. **`Asset`** specifies the one of the assets of the AMM instance. 
7. **`Asset2`** specifies the other asset of the AMM instance.

>    

Reviewer: There is a new field `OwnerNode`. It's a `page` hint for the `DirectoryNode` entry to link the account root and AMM object.

| --------------- | :----------------: | :-------: | :-----------: |
| `Amount2` | :heavy_check_mark: | `string` or `object`  |   `AMOUNT`    |

`Amount2` specifies the other pool asset of the AMM instance.

Both `Amount` and `Amount2` that represent issued assets MUST have `value` subfields specified.
        
---
        
| Field Name   |     Required?      | JSON Type | Internal Type |
| ------------ | :----------------: | :-------: | :-----------: |
| `TradingFee` | :heavy_check_mark: | `number` |   `UINT16`    |

`TradingFee` specifies the fee, in basis point, to be charged to the traders for the trades executed against the AMM instance. Trading fee is a percentage of the trading volume. Valid values for this field are between 0 and 1000 inclusive. A value of 1 is equivalent to 1/10 bps or 0.001%, allowing trading fee between 0% and 1%. 

---

The `AMMCreate` transaction MUST fail if the account issuing this transaction:

i. does not have sufficient balances, OR

ii. is NOT the `issuer` of either token AND the `RequireAuth` flag for the corresponding token is set AND the account inititaing the transaction is not authorized.

If the transaction is successful, 

> i. Two new ledger entries **`AccountRoot`** and `AMM` are created.

Reviewer: Three new ledger entries are created. `DirectoryNode` entry is the third one.

  - issuer: AMM instance `AccountID` and;
  - currency: the currency code for `LPTokens` for currency codes cur1 and cur2 is formed deterministically as follows:

  - Compute `SHA256{cur1, cur2}` 
  - `LPTokenID` = 0x03 + first 19 bytes from SHA256

The prefix 0x03 is added to identify `LPTokens`

The `value` field is computed as follows:

$$LPTokens = \Gamma_{A}^{W_{A}} *  \Gamma_{B}^{W_{B}} $$

where,

- $\Gamma_{A}$: Balance of asset $A$
- $\Gamma_{B}$: Balance of asset $B$
- $W_{A}$ & $W_{B}$: Normalized weights of asset $A$ and asset $B$ in the pool respectively

Initially by default, $W_{A}$ = $W_{B}$ = 0.5

<NOTE to self: Address the cases when authorization behaviours change on a trustline.>

#### 2.2.3. Cost of creating AMM instance `AccountRoot` and `AMM` ledger entries

Unlike other objects in the XRPL, there is no reserve for the `AccountRoot` and `AMM` ledger entries created by an `AMMCreate` transaction. Instead there is a higher `Fee` (~ 1 reserve) for `AMMCreate` transaction in XRP which is burned as a special transaction cost.

### 2.3. `Deleting the AMM instance `AccountRoot` and `AMM` ledger entries`

>  On final withdraw (i.e. when `LPTokens` balance is 0) the AMM instance automatically deletes upto 512 trustlines. If there are fewer then 512 trustlines then `AMM` object and `AccountRoot` objects are deleted.  

Reviewer: The `DirectoryNode` entry is also deleted.

>  However, if there are more than 512 trustlines then AMM instance is in empty state (`LPTokens` balance is 0) and we introduce a new transaction type `AMMDelete` to delete the remaining trustlines. .  Any one can call `AMMDelete` transaction. `AMMDelete` also has a limit of 512 trustlines and deletes the `AMM` object and `AccountRoot` objects only if there are fewer than 512 trustlines. If there are more trustlines to delete then `AMMDelete` returns `tecINCOMPLETE` code and user should submit `AMMDelete` again.

Reviewer: Typo - extra dot.

>  However, if there are more than 512 trustlines then AMM instance is in empty state (`LPTokens` balance is 0) and we introduce a new transaction type `AMMDelete` to delete the remaining trustlines. .  Any one can call `AMMDelete` transaction. `AMMDelete` also has a limit of 512 trustlines and deletes the `AMM` object and `AccountRoot` objects only if there are fewer than 512 trustlines. If there are more trustlines to delete then `AMMDelete` returns `tecINCOMPLETE` code and user should submit `AMMDelete` again.

Reviewer: Also `DirectoryNode` entry is deleted.

Author: Added the sentence to mention this.

---

| Field Name       | Required? | JSON Type | Internal Type |
| ---------------- | :-------: | :-------: | :-----------: |
| `Amount2` |           | `string` or `object` |   `AMOUNT`    |

`Amount2` specifies the details of other pool asset that the trader is willing to add.

---

| Field Name | Required? | JSON Type | Internal Type |
| ---------- | :-------: | :-------: | :-----------: |
| `EPrice`    |           | `string`  |   `AMOUNT`    |

`EPrice` specifies the effective-price of the token out after successful execution of the transaction. For **`AMMDeposit`** transaction, the token out is always `LPToken`. 

Note that the relative pricing does not change in case of **all-asset** deposit transaction.
`EPrice` is an invalid field for all assets deposits. It should only be specified in case of single sided deposits.

---

| Field Name | Required? | JSON Type | Internal Type |
| ---------- | :-------: | :-------: | :-----------: |
| `LPTokenOut` |           | `string`  |   `AMOUNT`    |

`LPTokenOut` specifies the amount of shares of the AMM instance pools.

> 

Reviewer: Could also have `TradingFee`, which is valid if the option flag is `tfTwoAssetIfEmpty`.

Author: Added.

| `tfSingleAsset`      | 0x00080000  |   If set, it indicates `Amount` parameter and may optionally include `LPTokenOut` parameter|
| `tfTwoAsset`      | 0x00100000  |   If set, it indicates `Amount` and `Amount2` parameter combination and may optionally include `LPTokenOut` paramater|
 | `tfOneAssetLPToken`      | 0x00200000 |   If set, it indicates `Amount` and `LPTokenOut` parameter combination|
 | `tfLimitLPToken`      | 0x00400000  |   If set, it indicates `Amount` and `EPrice` parameter combination|
 | `tfWithdrawAll`      | 0x00020000  |   If set, it indicates proportional withdrawal of all `LPTokens`|
| `tfTwoAssetIfEmpty`| 0x00800000  |   If set, it indicates that this deposit can only be submitted on an empty state AMM (LPTokens == 0)|

If `tfTwoAssetIfEmpty` is set, both amounts have to be specified  and deposited into AMM as is. It is sort of like `AMMCreate` in an empty AMM state.
 
 Following are the recommended valid combinations. Other invalid combinations may result in the failure of transaction.

- `LPTokenOut`, `[Amount]`, `[Amount2]`
- `Amount`, `[LPTokenOut]`

> - `Amount` , `Amount2`, `[LPTokenOut]`

Reviewer: Could have optional `TradingFee`, which is only valid with `tfTwoAssetIfEmpty` option.

Author: Added.

If the asset to be deposited is a token, specifying the `value` field is required, else the transaction will fail.

`LPTokenOut`, if included, specified minimum `LPTokens` amount that the user receives, else the transaction will fail.

3. Fields specified: `Amount`, `Amount2`, `[LPTokenOut]` and Flag: `tfTwoAsset`

Such a transaction assumes proportional deposit of pool assets with the constraints on the maximum amount of each asset that the trader is willing to deposit.

`LPTokenOut`, if included, specified minimum `LPTokens` amount that the user receives, else the transaction will fail.

4. Fields specified: `Amount` and `LPTokenOut` and Flag: `tfOneAssetLPToken`

Such a transaction assumes that a single asset `Amount` is deposited to obtain some share of the AMM instance's pools represented by amount of `LPTokenOut`. Since adding liquidity to the pool with one asset changes the ratio of the assets in the two pools, thus changing the relative pricing, trading fee is charged only on the amount of the deposited asset that causes this change.

5. Fields specified: `Amount` and `EPrice` and Flag: `tfLimitLPToken`

> 

Reviewer: Add 6. `Amount`, `Amount2`, [`TradingFee`] and Flag: `tfTwoAssetIfEmpty`.

Author: Added

#### 4.1.2. `AuctionSlot` field
We introduce a new object field `AuctionSlot` in the **`AMM`** object associated with each AMM instance. The `AuctionSlot` field has the following subfields: 

| Field Name | Required? | JSON Type | Internal Type |
| ---------- | :-------: | :-------: | :-----------: |
| `Account`    | :heavy_check_mark:          | `string` |   `AccountID`    |

`Account` represents the XRPL account that currently owns the auction slot.
        
---

| Field Name | Required? | JSON Type | Internal Type |
| ---------- | :-------: | :-------: | :-----------: |
| `Expiration`    | :heavy_check_mark:          | `string` |   `UINT32`    |

`Expiration` represents the number of seconds since the  [Ripple Epoch](https://xrpl.org/basic-data-types.html#specifying-time). This marks the end of the time from when slot was bought.
        
---

| Field Name | Required? | JSON Type | Internal Type |
| ---------- | :-------: | :-------: | :-----------: |
| `DiscountedFee`    |  :heavy_check_mark:         | `string` |   `UINT32`    |

`DiscountedFee` represents the `TradingFee` to be charged to this account for trading against the AMM instance. By default it is 0.

>         

Reviewer: `DiscountedFee` is `TradingFee/10`.

Author: Done

`AuthAccounts` represents an array of XRPL account IDs that are authorized to trade at the discounted fee against the AMM instance. The proposal allows for up to a maximum of four accounts.

---

| Field Name | Required? | JSON Type | Internal Type |
| ---------- | :-------: | :-------: | :-----------: |
| `MaxBidPrice`    |           | `string` |   `STRING NUMBER`    |

`MaxBidPrice` represents the maximum price that the bidder wants to pay for the slot. It is specified in units of `LPTokens`. This is not a required field. If specified let `MaxBidPrice` be `X` and let the slot-price computed by price scheduling slgorithm be `Y`, then bidder always pays the min(X, Y).     

---

## Appendices
### Appendix A. Specifying different parameters for **`AMMDeposit`** transaction

The proposal allows for traders to specify different combinations of the fields for **`AMMDeposit`** transaction. The implementation will figure out the best possible sub operations based on trader's specifications. Here are the recommended valid combinations. Other invalid combinations may result in the failure of transaction.

- `LPTokenOut`, `[Amount]`, `[Amount2]`
- `Amount`, `[LPTokenOut]`

> - `Amount` , `Amount2`, `[LPTokenOut]`

Reviewer: Add optional `TradingFee` (`tfTwoAssetsIfEmpty`).

Author: Done.

- Use equation 1 to compute the amount of $\Delta_{LPTokenOut}$, given the amount in `Amount`. Let this be `Z`
- Use equation 2 to compute the amount of asset2, given $\Delta_{LPTokenOut}$  ~ `Z`. Let the computed amount of asset2 be `X`.
  - If `X` <= amount in `Amount2`:
    - The amount of asset1 to be deposited is the one specified in `Amount`
    - The amount of asset2 to be deposited is `X`
    - The amount of `LPTokenOut` to be issued is `Z`
  - If X > amount in `Amount2`:
    - Use equation 2 to compute $\Delta_{LPTokenOut}$, given the amount in `Amount2`. Let this be `W`
    - Use equation 1 to compute the amount of asset1, given $\Delta_{LPTokenOut}$ ~ `W` from above. Let the computed amount of asset1 be `Y`
    - If `Y` <= amount in `Amount`:
      - The amount of asset1 to be deposited is `Y`
      - The amount of asset2 to be deposited is the one specified in `Amount2`
      - The amount of `LPTokenOut` to be issued is `W`

else, failed transaction.

If `LPTokenOut` field exists, and the amount of `LPTokenOut` computed above is less than the that specified in this field, then the transaction fails.

> 

Reviewer: `Amount`, `Amount2`, [`TradingFee`] for `tfTwoAssetsIfEmpty`. This option is equivalent to `AMMCreate`, i.e. the pool is created with `Amount` and `Amount2` balances.

Author: Added this as 6th combination in the AMMDeposit section.

<pre>    
  Title:        <b>Automated Market Maker on XRPL</b>
  Revision:     <b>4</b> (2023-08-08)

<hr>  Author:       <a href="mailto:amalhotra@ripple.com">Aanchal Malhotra</a>
                <a href="mailto:david@ripple.com">David J. Schwartz</a>
  Affiliation:  <a href="https://ripple.com">Ripple</a>
</pre>
# Automated Market Maker on XRP Ledger

### Abstract

> The XRPL decentralized exchange (DeX) currently provides liquidity exclusively by manual market making and order books. This proposal adds non-custodial automated market maker (AMM) as a native feature to the XRPL DeX in a way that provides increased returns to those who provide liquidity for the AMM and minimizes the risk of losses due to volatility.

Reviewer 2: ```suggestion
The XRPL decentralized exchange (DEX) currently provides liquidity exclusively by manual market making and order books. This proposal adds non-custodial automated market maker (AMM) as a native feature to the XRPL DEX in a way that provides increased returns to those who provide liquidity for the AMM and minimizes the risk of losses due to volatility.
```
For clarity, it should be capitalized as DEX, as the E stands for Exchange.

![Untitled](https://github.com/XRPLF/XRPL-Standards/assets/81505/ff87686c-1ab5-4db7-a322-01a37d37807d)
![Untitled 2](https://github.com/XRPLF/XRPL-Standards/assets/81505/30548054-21d0-47e5-9905-66a293ffcdb8)
![Untitled 3](https://github.com/XRPLF/XRPL-Standards/assets/81505/1bd6d66c-e17a-4987-84a3-85fe0872fc97)

We propose a unique mechanism where the AMM instance continuously auctions off trading advantages to arbitrageurs, charging them and giving these earnings to its liquidity providers. This allows liquidity providers to take a large share of the profits that would normally be taken by arbitrageurs.

The AMM instance charges a spread on the trades that change the ratio of tokens in the instance's pools. This trading fee is added to the AMM instance's capital pool, thus adding to the liquidity providers' returns.

The AMM instances also provide governance rights to the largest share holders of the AMM instance. This allows them to vote on the trading fee the instance charges.

> XRPL's AMM based DeX interacts with XRPL's limit order book (LOB) based DeX so that users of AMM pools have access to all order flow and liquidity on LOB DeX, and vice versa. The payment and order placement transactors automatically determine whether swapping within a liquidity pool or through the order book will provide the best price for the user and execute accordingly. Pathfinding considers paths with both order books and AMMs in various combinations to improve the overall exchange rate.

Reviewer 2: ```suggestion
XRPL's AMM based DEX interacts with XRPL's limit order book (LOB)-based DEX so that users of AMM pools have access to all order flow and liquidity on LOB DEX, and vice versa. The payment and order placement transactors automatically determine whether swapping within a liquidity pool or through the order book will provide the best price for the user and execute accordingly. Pathfinding considers paths with both order books and AMMs in various combinations to improve the overall exchange rate.
```

### 1. Introduction

> AMMs are agents that pool liquidity and make it available to traders according to a predetermined algorithm. This is a proposal for geometric mean market maker (GM3) based DeX as a native XRPL feature. GM3 algorithmically discovers a fair exchange value, stabilized by arbitrage, to automatically produce liquidity. On XRPL, this would provide liquidity pools between XRP and issued assets as well as between any two issued assets. Arbitrageurs play a significant role in keeping the AMM DeX in a stable state with respect to external markets.

Reviewer 2: ```suggestion
AMMs are agents that pool liquidity and make it available to traders according to a predetermined algorithm. This is a proposal for geometric mean market maker (GM3)-based DEX as a native XRPL feature. GM3 algorithmically discovers a fair exchange value, stabilized by arbitrage, to automatically produce liquidity. On XRPL, this would provide liquidity pools between XRP and issued assets as well as between any two issued assets. Arbitrageurs play a significant role in keeping the AMM DEX in a stable state with respect to external markets.
```

Additionally, we introduce a new field called `AMMID` to the `AccountRoot` object. It replaces directory entry to link account root and AMM object. 

| Field Name        |     Required?      | JSON Type | Internal Type |
| ----------------- | :----------------: | :-------: | :-----------: |
| `AMMID` | :heavy_check_mark: | `string`  |   `UINT256`    |

`AMMID` specifies the `AMMID` as described below.

The unique ID of the **`AccountRoot`** object, a.k.a. **`AccountRootID`** is computed as follows:
- for (i = 0; i <= 256; i--)
  - Compute `AccountRootID` = `SHA512-Half`(i || [Parent Ledger Hash](https://xrpl.org/ledgerhashes.html) || `AMMID`)
  - If the computed `AccountRootID` exists, repeat
    - else, return `AccountRootID`
 
#### **`AMM`** ledger entry 
The unique ID of the new **`AMM`** object , a.k.a **`AMMID`** is computed as follows:
- Calculate the `SHA512-Half` of some of the following values:
  - The `issuer` of the issued asset;
  - The `currency` of the issued asset;
  - The `issuer` of the second issued asset, if there exists one;
  - The `currency` of the second issued asset, if there exists one;

>   <!-- `XRP` if the `Balance` field exists, i.e. if one of the assets is XRP "GT: why is this needed? We just use issuer/currency of each asset. XRP defaults both to 0."-->

Reviewer 2: The serialization format is documented here: https://xrpl.org/serialization.html#currency-codes

> The currency code `0x0000000000000000000000000000000000000000` (all zeroes) is generally disallowed. Usually, XRP amounts are not specified with currency codes. However, this code is used to indicate XRP in rare cases where a field must specify a currency code for XRP.

AND

$$\Delta_{B} = \left(\frac{\Delta_{LPTokens}}{\Gamma_{LPTokens}}\right) * \Gamma_{B} \tag{2}$$

Following is the updated pool composition of the AMM instance after successful transaction:

- $\Gamma_{A} = \Gamma_{A} + \Delta_{A}$: Current new balance of asset $A$
- $\Gamma_{B} = \Gamma_{B} + \Delta_{B}$: Current new balance of asset $B$
- $\Gamma_{LPTokens} = \Gamma_{LPTokens} + \Delta_{LPTokens}$: Current new balance of outstanding `LPTokens`

#### 2.4.1.3 Single asset deposit

Let $B$ be the only asset being deposited in return for $\Delta_{LPTokens}$, then this is executed as a combination of _equal asset_ deposit and a _swap_ transaction. The trading fee is only charged on the _amount_ of asset being swapped.

Also let 

- $F_{1} = (1-TFee)$
- $F_{2} = \frac{(1 - \frac{TFee}{2})}{F_1}$, then

$$\Delta_{LPTokens} = \Gamma_{LPTokens} * \left[\frac{\frac{\Delta_{B}}{\Gamma_{B}} - \left[\left({F_2}^{\frac{1}{W_{B}}} + \frac{\Delta_{B}}{(\Gamma_{B}*F_1)}\right)^{W_B} - F_2\right]} 
        {1 + {\left(F_2^{\frac{1}{W_B}} + \frac{\Delta_{B}}{(\Gamma_{B}*F_1)}\right)}^{W_{B}} - F_2}\right] \tag{3}$$

> Similarly,
> Similarly, we can derive $\Delta{B}$ from the above equation. We call this equation 4.

Reviewer 2: ```suggestion
Similarly, we can derive $\Delta{B}$ from the above equation. We call this equation 4.
```
Just removed extra "Similarly,"

  - issuer: AMM instance `AccountID` and;
  - currency: the currency code for `LPTokens` for currency codes cur1 and cur2 is formed deterministically as follows:

  - Compute `SHA256{cur1, cur2}` 
  - `LPTokenID` = 0x03 + first 19 bytes from SHA256

The prefix 0x03 is added to identify `LPTokens`

The `value` field is computed as follows:

$$LPTokens = \Gamma_{A}^{W_{A}} *  \Gamma_{B}^{W_{B}} $$

where,

- $\Gamma_{A}$: Balance of asset $A$
- $\Gamma_{B}$: Balance of asset $B$
- $W_{A}$ & $W_{B}$: Normalized weights of asset $A$ and asset $B$ in the pool respectively

Initially by default, $W_{A}$ = $W_{B}$ = 0.5

<NOTE to self: Address the cases when authorization behaviours change on a trustline.>

#### 2.2.3. Cost of creating AMM instance `AccountRoot` and `AMM` ledger entries

Unlike other objects in the XRPL, there is no reserve for the `AccountRoot` and `AMM` ledger entries created by an `AMMCreate` transaction. Instead there is a higher `Fee` (~ 1 reserve) for `AMMCreate` transaction in XRP which is burned as a special transaction cost.

> ### 2.3. `Deleting the AMM instance `AccountRoot`, `AMM` and `DirectoryNode` ledger entries`

Reviewer 2: ```suggestion
### 2.3. Deleting the AMM instance `AccountRoot`, `AMM` and `DirectoryNode` ledger entries
```

On final withdraw (i.e. when `LPTokens` balance is 0) the AMM instance automatically deletes up to 512 trust lines. If there are fewer then 512 trustlines then `AMM`, `AccountRoot` and `DirecotoryNode` objects are deleted.  
 

> However, if there are more than 512 trustlines then AMM instance is in empty state (`LPTokens` balance is 0) and we introduce a new transaction type `AMMDelete` to delete the remaining trustlines. Any one can call `AMMDelete` transaction. `AMMDelete` also has a limit of 512 trustlines and deletes the `AMM`, `AccountRoot` and `DirectoryNode` objects only if there are fewer than 512 trustlines. If there are more trustlines to delete then `AMMDelete` returns `tecINCOMPLETE` code and user should submit `AMMDelete` again.

Reviewer 2: ```suggestion
However, if there are more than 512 trustlines then AMM instance remains in empty state (`LPTokens` balance is 0). To handle cleaning this up, we introduce a new transaction type `AMMDelete` to delete the remaining trustlines. Anyone can call the `AMMDelete` transactor. `AMMDelete` also has a limit of 512 trustlines and deletes the `AMM`, `AccountRoot` and `DirectoryNode` objects only if there are fewer than 512 trustlines at the time it executes. If there are more trustlines to delete, then `AMMDelete` returns the `tecINCOMPLETE` result code and the user should submit another `AMMDelete` to delete more entries.
```

b. The effective price of asset traded out does not exceed the amount specified in `EPrice`

- Use equations 8 & `III` and amount in `EPrice`to compute the two variables, i.e., 
  - asset in as `LPTokenIn`. Let this be `X`
  - asset out as that in `Amount`. Let this be `Y` 
- If (amount exists for `Amount` & `Y` >= amount in `Amount`) || (amount field does not exist for `Amount`):
  - The amount of assetOut is given by `Y`
  - The amount of LPTokens is given by `X`

else transaction fails.

Following updates after a successful **`AMMWithdraw`** transaction:

- The withdrawn asset, if XRP, is transferred from AMM instance account to the account that initiated the transaction, thus changing the `Balance` field of each account
- The withdrawn asset, if token, is balanced between the AMM instance account and the issuer account.
- The `LPTokens` ~ $\Delta_{LPTokenIn}$ are balanced between the AMM instance account and the account that initiated the transaction.
- The pool composition is updated. Conservation function is not preserved in case of liquidity withdrawal (and is not expected to.)

## Implementation

> A proposed reference implementation is in progress here: https://github.com/XRPLF/rippled/pull/4294

Reviewer 2: ```suggestion
The reference implementation of this spec can be found in `rippled`:

- https://github.com/XRPLF/rippled/pull/4294
- https://github.com/XRPLF/rippled/pull/4626
- https://github.com/XRPLF/rippled/pull/4674
- https://github.com/XRPLF/rippled/pull/4682
```

##### **`AccountRoot`** ledger entry

> We introduce a new field called `AMMID` to the `AccountRoot` object. It replaces directory entry to link account root and AMM object. 

Reviewer: It doesn't replace the directory entry. We still have the directory entry. It does provide an easy link to AMM object and a way to identify AMM AccountRoot.

Author: Fixed

Unlike other objects in the XRPL, there is no reserve for the `AccountRoot` and `AMM` ledger entries created by an `AMMCreate` transaction. Instead there is a higher `Fee` (~ 1 reserve) for `AMMCreate` transaction in XRP which is burned as a special transaction cost.

### 2.3. `Deleting the AMM instance `AccountRoot`, `AMM` and `DirectoryNode` ledger entries`
On final withdraw (i.e. when `LPTokens` balance is 0) the AMM instance automatically deletes up to 512 trust lines. If there are fewer then 512 trustlines then `AMM`, `AccountRoot` and `DirecotoryNode` objects are deleted.  
 

> However, if there are more than 512 trustlines then AMM instance is in empty state (`LPTokens` balance is 0) and we introduce a new transaction type `AMMDelete` to delete the remaining trustlines. Any one can call `AMMDelete` transaction. `AMMDelete` also has a limit of 512 trustlines and deletes the `AMM` object and `AccountRoot` objects only if there are fewer than 512 trustlines. If there are more trustlines to delete then `AMMDelete` returns `tecINCOMPLETE` code and user should submit `AMMDelete` again.

Reviewer: "`AMMDelete` also has a limit of 512 trustlines and deletes `AMM`, `AccountRoot`, and `DirectoryNode` objects..."

Author: Fixed.
